overlord/exportstate/exportstate.go

```diff
@@ -0,0 +1,144 @@
+	var exports map[string]*json.RawMessage
+	if err := st.Get("exports", &exports); err != nil && err != state.ErrNoState {
+		panic("internal error: cannot unmarshal exports state: " + err.Error())
+	}
+	if exports == nil {
+		exports = make(map[string]*json.RawMessage)
+	}
+	key := stateMapKey(instanceName, rev)
+	if m == nil {
+		delete(exports, key)
+	} else {
+		data, err := json.Marshal(m)
+		if err != nil {
+			panic("internal error: cannot marshal export manifest: " + err.Error())
+		}
+		raw := json.RawMessage(data)
+		exports[key] = &raw
+	}
+	st.Set("exports", exports)
+}
+
+// Get recalls export manifest of a particular snap revision.
+func Get(st *state.State, instanceName string, rev snap.Revision, m *Manifest) error {
+	*m = Manifest{}
+
+	var exports map[string]*json.RawMessage
+	if err := st.Get("exports", &exports); err != nil {
+		// This can return state.ErrNoState
+		return err
+	}
+	key := stateMapKey(instanceName, rev)
+	raw, ok := exports[key]
+	if !ok {
+		return state.ErrNoState
+	}
+	// XXX: do we need the address?
+	if err := json.Unmarshal([]byte(*raw), &m); err != nil {
+		return fmt.Errorf("cannot unmarshal export manifest: %v", err)
+	}
+	return nil
+}
+
+// exporetedVersionSymlinkPath returns the path of the current exported version symlink
+// for given export name.
+func exportedVersionSymlinkPath(exportedName string) string {
+	return filepath.Join(dirs.ExportDir, exportedName, "current")
+}
+
+// updateExportedVersion updates or removes the exported version symlink.
+//
+// If exportedVersion is not empty then it indicates the new version to use.
+// If the symbolic link cannot be created because the export directory does not
+// exist no error is reported. This is because this function is most often
+// called from link-snap, where it runs unconditionally, but most snaps do not
+// have any content to export and the symlink would be dangling.
+//
+// if exportedVersion is empty then then it indicates that no version is exported
+//
+// Appropriate version can be computed by exportedVersionForSnap.
+func updateExportedVersion(exportedName, exportedVersion string) error {
+	pathName := exportedVersionSymlinkPath(exportedName)
+	if exportedVersion != "" {
+		if err := osutil.AtomicSymlink(exportedVersion, pathName); err != nil && !os.IsNotExist(err) {
+			return fmt.Errorf("cannot set exported version of %q to %q: %v", exportedName, exportedVersion, err)
+		}
+	} else {
+		if err := os.Remove(pathName); err != nil && !os.IsNotExist(err) {
+			return fmt.Errorf("cannot unset exported version of %q: %v", exportedName, err)
+		}
+	}
+	return nil
+
+}
+
+// ExportedNameVersion returns the (snapName, exportedVersion) tuple to
+// use as the current provider of all the export sets of a given snap. The
+// returned exportedVersion may be empty, indicating that given snap has no
+// current revision.
+func ExportedNameVersion(st *state.State, instanceName string) (exportedName string, exportedVersion string, err error) {
```

Comment:
maybe it made more sense originally but in the simplified world it is strange that we use "exported name" for the name of the snap that is a source of export,  it would seem more appropriate for the names of things inside a export set.
 
Maybe we should change we call it "source" and "source name" and change the emphasis here,  by switching things:

ExportedVersionAndSource  and swapping the results, anyway in most cases exportedName == instanceName.

Reply:
Yeah, I think this is sensible. I've swapped the arguments around here and in the place where we use the return values `updateExportedVersion`.